[FLINK-2732] Display TM logs in Dashboard

zentol opened the pull request:
This PR allows the TM logs to be displayed in the Dashboard.

The page displaying information about a single TaskManager now has two additional tabs (previously only metrics), log and stdout, similarly to the JobManager page.

When this tabs are accessed, a TaskManagerLogHandler is called that roughly does the following:
* get the TM ID from the REST API call
* ask JM for the TM gateway belonging to this ID
* tell the TM to upload his logs to the Blob store, and to return the BlobKey
* send the BlobKey to the JM , asking for the path to the log file
* display file in the Dashboard

Should the BlobService not be available (e.g when the start-local script is used), a proper message is displayed instead of the log.

At all times only a single version of the TM log/out file exists in the Blob store; whenever the log is requested the old entry is deleted, if the file has changed. This is handled by the TaskManagerLogHandler. To prevent race-conditions when spamming the refresh button, only a single transfer can be active for each TM. Should the user hit refresh while a transfer is in progress, "loading..." will be displayed instead of the log.

When the TM uploads it's log a separate Thread is created to do that, so that the TM isn't blocked for a while if the file is too big. Similarly, the handler doesn't block while waiting the the TM upload, and instead uses a FutureCallback to handle the finished upload.

uce commented:
Thanks for the PR and the description. Could you add screenshots so that more people can get a quick idea of the change? :)

zentol commented:
Screenshots added:
![TaskManager main page](https://cloud.githubusercontent.com/assets/5725237/13760504/c270cafc-ea33-11e5-9ec8-3c86ceb8367a.png)
![lTaskManager log page](https://cloud.githubusercontent.com/assets/5725237/13760505/c271649e-ea33-11e5-945b-1ba2f4d70f9c.png)
![s3](https://cloud.githubusercontent.com/assets/5725237/13760572/1d124436-ea34-11e5-9098-60f634d752b4.png)

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+ *
+ * See http://netty.io and
+ * https://github.com/netty/netty/blob/4.0/example/src/main/java/io/netty/example/http/file/HttpStaticFileServerHandler.java
+ *****************************************************************************/
```
Good attribution :+1:

rmetzger commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+public class TaskManagerLogHandler extends SimpleChannelInboundHandler<Routed> {
+
+	/** Default logger, if none is specified */
+	private static final Logger DEFAULT_LOGGER = LoggerFactory.getLogger(StaticFileServerHandler.class);
```
The logger should probably be of `TaskManagerLogHandler`.

rmetzger commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+							lastContentFuture.addListener(ChannelFutureListener.CLOSE);
+						}
+					} catch (Exception e) {
+						logger.error("Serving of taskmanager log failed: " + e.getMessage());
```
I would add the exception to the log, not only the message

rmetzger commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -781,6 +787,32 @@ class TaskManager(
+        override def run(): Unit = {
+          val jmHost = jobManager.path.address.host.get
+          val port = blobService.get.getPort
+          val logFilePath = System.getProperty("log.file");
```
I don't think we can rely on the system property to be set correctly.
I think the log directory is in some configuration

zentol replied:
i looked through the ConfigConstants class and couldn't find a key specific to the log files, as in equivalent to JOB_MANAGER_WEB_LOG_PATH_KEY.

rmetzger replied:
Okay. It seems that the JobManager is trying to get it from the property. if that is not set, it tries to get it from the config and then it fails.

I think we should do something similar for the TaskManager as well. This allows users running into issues with this to configure a path.

zentol replied:
and how would we pass that value to the logger?

rmetzger commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -781,6 +787,32 @@ class TaskManager(
+    if (blobService.isDefined && jobManager.path.address.host.isDefined) {
+      //create new Thread to upload log, to not block the TM for too long
+      new Thread() {
+        override def run(): Unit = {
```
Exceptions thrown from this run() method are not properly catched.
Any failures in the thread will be written to system out.

tillrohrmann replied:
Should be system error to where it is written.

rmetzger commented:
I think we should an automated test for the change.
The `WebFrontendITCase` or one of the YARN tests is probably the right place for such a test.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/messages/JobManagerMessages.scala:
```diff
@@ -460,6 +461,10 @@ object JobManagerMessages {
 
+  case class RequestTaskManagerLog(blobKey: BlobKey)
+
+  case class DeleteTaskManagerLog(blobKey: BlobKey)
```
ScalaDocs missing

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/messages/JobManagerMessages.scala:
```diff
@@ -460,6 +461,10 @@ object JobManagerMessages {
     */
   case class DisposeSavepointFailure(cause: Throwable)
 
+  case class RequestTaskManagerLog(blobKey: BlobKey)
```
ScalaDocs missing

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/messages/TaskManagerMessages.scala:
```diff
@@ -114,6 +114,16 @@ object TaskManagerMessages {
+
+  case class RequestTaskManagerLog(requestType : LogTypeRequest)
+
+  case class IsBlobServiceDefined()
```
should be a case object

tillrohrmann replied:
ScalaDocs missing

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -781,6 +787,32 @@ class TaskManager(
 
+  private def handleRequestTaskManagerLog
+  (sender: ActorRef, requestType: LogTypeRequest, jobManager: ActorRef):
+  Unit = {
```
Formatting is off here. I would suggest

```
private def xyz(
    a: A,
    b: B)
  : C = {
}
```

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -781,6 +787,32 @@ class TaskManager(
+
+          sender ! client.put(new FileInputStream(file))
+        }
+      }.run()
```
This won't start a new thread but simply call the just overwritten `run` method. Use `start` instead. Furthermore, it is not encouraged to extend the Thread class. It is better to implement a `Runnable` or even better avoiding to deal with raw threads at all. I guess it would be easier to simply spawn a `Future` which is executed by a respective `ExecutionContext`.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -781,6 +787,32 @@ class TaskManager(
+        }
+      }.run()
+    } else {
+      throw new IOException("No BlobService defined, cannot upload TaskManager logs.")
```
Throwing this exception here will terminate the TM. I think the exception should rather be send to the `sender` where it is logged.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -781,6 +787,32 @@ class TaskManager(
+  private def handleRequestTaskManagerLog
+  (sender: ActorRef, requestType: LogTypeRequest, jobManager: ActorRef):
+  Unit = {
+    if (blobService.isDefined && jobManager.path.address.host.isDefined) {
```
If the `blobService` is defined the `jobManager` should have a defined address as well. I think it would be better to use pattern matching to extract the blobService from the `Option`. This is more scalaesque.

zentol replied:
this check is made specifically for the start-local.sh script, for which the jobManager does not have a defined address.

tillrohrmann replied:
Maybe we could extend the `BlobService` so that it can create a `BlobClient` for you. Then you don't have to worry about the address, because the `BlobService` should know it.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -781,6 +787,32 @@ class TaskManager(
+      new Thread() {
+        override def run(): Unit = {
+          val jmHost = jobManager.path.address.host.get
+          val port = blobService.get.getPort
```
Maybe we should extend the `BlobService` to return the `serverAddress` and not only the port.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -310,6 +310,12 @@ class TaskManager(
+      handleRequestTaskManagerLog(sender(), requestType, currentJobManager.get)
+
+    case IsBlobServiceDefined() =>
+      sender() ! (blobService.isDefined && currentJobManager.get.path.address.host.isDefined)
```
if the `blobService` is defined, then it should have a valid `serverAddress` where the `BlobServer` is running.

zentol replied:
this is not the case when using the start-local.sh script.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -310,6 +310,12 @@ class TaskManager(
       killTaskManagerFatal(message, cause)
+
+    case RequestTaskManagerLog(requestType : LogTypeRequest) =>
+      handleRequestTaskManagerLog(sender(), requestType, currentJobManager.get)
```
Why not simply returning a `Future[BlobKey]` from the `handleRequestTaskManagerLog` which is simply piped to the `sender`?

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+ * example.</p>
+ */
+@ChannelHandler.Sharable
+public class TaskManagerLogHandler extends SimpleChannelInboundHandler<Routed> {
```
Why do we implement a new `SimpleChannelInboundHandler` instead of a `RequestHandler`. That latter way would avoid a considerable amount of redundant code.

zentol replied:
The RequestHandler expects a String to be returned. Do you think it's a good idea to do that since it means allocating a string containing the full log?

In addition, i wanted to reuse the code that is used to serve the jobmanager files, as i can't assess in general whether the Requesthandler exposes everything we need; can you?

tillrohrmann replied:
What do you need for that? I would assume that everything is already there.

What if we implement a different version of the `RequestHandler` which receives the `ChannelHandlerContext`. Or we create a `RuntimeMonitorHandlerBase` which implements `channelRead0` but not `respondLeader`.

zentol replied:
a good solution would be to extend the StaticFileServerHandler, and modify some fields/methods to be protected instead of final. this way we could remove as much redundant code as possible.

zentol replied:
only problem being that StaticFileServerhandler has a File constructor argument...

zentol replied:
going for the RuntimeMonitorhandlerBase approach.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+	private final Logger logger;
+
+	/** indicates which log file should be displayed; true indicates .log, false indicates .out */
+	private final boolean serveLogFile;
```
I think it would be better to use an enum here.

zentol replied:
i think it's overkill for a binary option, and won't make the code more readable.

tillrohrmann replied:
But it is not intuitive that if the `serveLogFile` is false, then the stdout file is served. I would expect that if `serveLogFile` is false, that the log file won't be served. This becomes obvious if you look at the instantiation of the `TaskManagerLogHandler` where you have an boolean value as parameter which one does not understand if you don't look in the code.

zentol replied:
i think i found a solution that keeps us both happy...

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+			}
+			fetch = lastRequestCompleted.get(taskManagerID);
+			lastRequestCompleted.put(taskManagerID, false);
+		}
```
I think it would be easier to use a `ConcurrentHashMap` in combination with `putIfAbsent` to guard against multiple concurrent requests.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+			//get taskmanager gateway
+			InstanceID instanceID = new InstanceID(StringUtils.hexStringToByte(taskManagerID));
+			Future<Object> future = jobManager.ask(new JobManagerMessages.RequestTaskManagerInstance(instanceID), timeout);
+			JobManagerMessages.TaskManagerInstance instance = (JobManagerMessages.TaskManagerInstance) Await.result(future, timeout);
```
This is a blocking operation which we want to avoid in the a request handler. But using future composition to do the job.

zentol replied:
so a triple nested future then?

tillrohrmann replied:
No simply use future composition: `map`, `zip`, etc. That way you can simply define operations for future values. See also http://doc.akka.io/docs/akka/snapshot/scala/futures.html

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+
+			Future<Object> isBlobServiceDefined = taskManager.getActorGateway().ask(TaskManagerMessages.getIsBlobServiceDefined(), timeout);
+
+			if (!(Boolean) Await.result(isBlobServiceDefined, timeout)) {
```
Same problem here with blocking operation.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+							}
+						} else {
+							(serveLogFile ? lastSubmittedLog : lastSubmittedStdout).put(taskManagerID, blobKey);
+						}
```
I think this code block would be a bit easier to read if we refrained from using too much nesting and ternary operators.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+
+						//send blobkey to jobmanager
+						Future<Object> logPathFuture = jobManager.ask(JobManagerMessages.getRequestTaskManagerLog(blobKey), timeout);
+						String filePath = (String) Await.result(logPathFuture, timeout);
```
Blocking operation. It would be better to use future composition here.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+				@Override
+				public void onComplete(Throwable failure, Object success) throws Throwable {
+					try {
+						BlobKey blobKey = (BlobKey) success;
```
`success` can also be null

zentol replied:
if success is null, is failure always != null?

tillrohrmann replied:
Not if the receiver sent `null` back.

tillrohrmann replied:
You should first check for the failure case and then handle the success case.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+			Future<Object> blobKeyFuture = taskManager.getActorGateway().ask(serveLogFile ? TaskManagerMessages.getRequestTaskManagerLog() : TaskManagerMessages.getRequestTaskManagerStdout(), timeout);
+			blobKeyFuture.onComplete(new OnComplete<Object>() {
+				@Override
+				public void onComplete(Throwable failure, Object success) throws Throwable {
```
What if the `failure != null`?

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+								@Override
+								public void operationComplete(io.netty.util.concurrent.Future<? super Void> future) throws Exception {
+									synchronized (lastRequestCompleted) {
+										lastRequestCompleted.put(taskManagerID, true);
```
What if a failure occurs before this listener is added. Wouldn't that mean that the `lastRequestCompleted` entry for this `taskManagerID` would remain `false` for all eternity?

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+			}, executor);
+		} else {
+			display(ctx, request, "loading...");
+			return;
```
not necessary

zentol replied:
what exactly?

tillrohrmann replied:
`return;`

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java:
```diff
@@ -0,0 +1,345 @@
+						}
+
+						//send blobkey to jobmanager
+						Future<Object> logPathFuture = jobManager.ask(JobManagerMessages.getRequestTaskManagerLog(blobKey), timeout);
```
I don't like that we request the file path here. This more or less means that the handler must have access to the filesystem as the JM has. In the long run, we want to disentangle the `WebRuntimeMonitor` from the `JobManager` so that both components are independent of each other. I would rather prefer that the `WebRuntimeMonitor` has its own `BlobService` which it uses to request the log file given the `BlobKey` received from the TaskManager. In the first implementation, this `BlobService` could simply be the `BlobServer` of the JM. By doing it this way, we would create an abstraction which would make it easier to separate the `WebRuntimeMonitor` from the `JobManager`.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java:
```diff
@@ -234,7 +235,9 @@ public WebRuntimeMonitor(
 			.GET("/taskmanagers", handler(new TaskManagersHandler(DEFAULT_REQUEST_TIMEOUT)))
-			.GET("/taskmanagers/:" + TaskManagersHandler.TASK_MANAGER_ID_KEY, handler(new TaskManagersHandler(DEFAULT_REQUEST_TIMEOUT)))
+			.GET("/taskmanagers/:" + TaskManagersHandler.TASK_MANAGER_ID_KEY + "/metrics", handler(new TaskManagersHandler(DEFAULT_REQUEST_TIMEOUT)))
+			.GET("/taskmanagers/:" + TaskManagersHandler.TASK_MANAGER_ID_KEY + "/log", new TaskManagerLogHandler(retriever, actorSystem.dispatcher(), jobManagerAddressPromise.future(), timeout, true))
```
I think we shouldn't use the dispatcher of the `ActorSystem` as the `ExecutorService` here. The `ActorSystem` is usually the `ActorSystem` of the JM and we shouldn't put load on its `ThreadPool`. I would rather use a dedicated `ThreadPool` for the execution of the futures.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/messages/TaskManagerMessages.scala:
```diff
@@ -114,6 +114,16 @@ object TaskManagerMessages {
+
+  case class RequestTaskManagerLog(requestType : LogTypeRequest)
+
+  case object IsBlobServiceDefined
```
ScalaDocs still missing

tillrohrmann commented:
I think the PR goes in the right direction @zentol. I had some comments concerning the concurrent implementation of the `TaskManagerLogHandler` and on the side of the `TaskManager`. These should be addressed before merging.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -310,6 +310,12 @@ class TaskManager(
+      handleRequestTaskManagerLog(sender(), requestType, currentJobManager.get)
+
+    case IsBlobServiceDefined =>
+      sender() ! (blobService.isDefined && currentJobManager.get.path.address.host.isDefined)
```
What is the value of `path` in the local case? Maybe we should then extend the `BlobService` to simply give back a `BlobClient` which has the right connection address.

zentol commented:
First batch of comments addressed, the following are still missing:
* test cases
* re-implement as something that is not a SimpleChannelInboundHandler
* possible abstraction to separate jobmanager/dashboard

tillrohrmann commented:
The jobmanager/dashboard separation should be really easy to do. You just have to give the `WebRuntimeMonitor` a `BlobService` as a constructor parameter. And now you can simply pass in the JobManager's `BlobServer` as a first implementation.

zentol commented:
yeah...the changes in the jobmanager are fairly easy, albeit hacky and ugly. the FlinkMiniCluster however is pain.

zentol commented:
@tillrohrmann i don't think we should do this now. We would have to change several method related to the creation of the JobManager (to return the BlobService), potentially hack around in the tests to get the BlobService from the active Jobmanager to feed it into the WebMonitor (which is created separately).

The gains are negligible; it a 2 minute fix to change the Handler itself.

zentol commented:
second batch of changes coming in.
* TaskManager checks whether blobService is defined
* new class: RuntimeMonitorHandlerBase
 * RuntimeMonitorHandler/TaskManagerLogHandler extend RuntimeMonnitorHandlerBase
 * removes some redundant code
* TaskManagerLogHandler uses it's own BlobCache
 * reverted addition of messages to JobManagerMessages
 * reverted addition of methods to LibraryCacheManager
 * new method in BlobCache, to delete a Blob both locally and on the server

Issues left:
* tests

zentol commented:
Third batch of changes:
* TaskManager log file location config key added
* Test added
* FileInputStream in handleRequestTaskManagerLog properly closed

I just ran out of issues to address.

rmetzger commented:
I just tested the change on a cluster, also with larger log files, and it seems to work very well.
The only thing I found was that clicking the "Download" button on the top right lead to the following error: "Failure: 404 Not Found"

rmetzger commented:
Also, I'm not sure if this log message is correct:
```
2016-03-18 12:59:43,376 INFO  org.apache.flink.runtime.blob.BlobCache                       - Downloading ee982985ecf8b6a9e3d33eb4eb7d84f480666e33 from localhost/127.0.0.1:55764
2016-03-18 12:59:48,594 INFO  org.apache.flink.runtime.blob.BlobCache                       - Downloading a375a60bb40c8f95c786bcffc77c7b7c15a9ddf3 from localhost/127.0.0.1:55764
2016-03-18 12:59:56,727 INFO  org.apache.flink.runtime.blob.BlobCache                       - Downloading 6077377ab5a146b9c2f0917d2e1fd03b8cfdbe84 from localhost/127.0.0.1:55764
```

It says its downloading the logs locally, but they are loaded from the remote TM, right?

zentol commented:
the files are currently loaded twice. First, the logs are copied from the Taskmanager to the JobManager blob store. They are then copied again from the JobManager blob store to the TaskmanagerLogHandler BlobCache. This was @tillrohrmann idea, and is meant to make the handler easier to change when the Dashboard and JobManager are separated.

zentol commented:
It also reduces the number of changed classes, since no new messsages to the jobmanager had to be added.

rmetzger commented:
I see

tillrohrmann commented:
We could also have passed the `BlobServer` of the `JobManager` to the `TaskManagerLogHandler` to avoid the extra copying. But this would mean that we would have to move the `WebRuntimeMonitor` creation in the `startJobManagerActors` method, which is not so nice.

zentol commented:
The download button should be functional now.

rmetzger commented:
Cool! Thank you for fixing it so quickly.

From my point of view, the change is good to merge. I've tested it on a cluster and locally.

mxm commented:
Looks good to merge. @zentol Could you rebase this to the latest master?

zentol commented:
Rebased.

zentol commented:
There appears to be a problem with the test, looking into it...

zentol commented:
Test failures should be resolved.

zentol commented:
Merging.
